CIPs/CIP-69/CIP-69.md

```diff
@@ -0,0 +1,108 @@
+---
+cip: 69
+title: Batched Anchor Data Structure
+author: Joel Thorstensson (@oed)
+discussions-to: https://github.com/ceramicnetwork/CIP/issues/69
+status: Last Call
+category: Standards
+type: Core
+created: 2020-11-02
+---
+
+## Simple Summary
+This CIP describes the data structure of a blockchain anchor that batches multiple document updates together. 
+
+
+## Abstract
+An anchor service accepts document update anchor request from clients and batches them together in a single merkle tree, which is then anchored into a specific blockchain. The anchor service also sorts all of the requests and creates a bloom filter which describes which documents where updated in this batch, which is also included in the blockchain transaction. This allows external observers to more effectively determine if a document which they are interested in was updated.
+
+
+## Motivation
+There are two main motivations for this CIP. Firstly having an agreed upon spec for the structure of the batched anchor merkle tree makes it easier for all consumers of the ceramic protocol to know what to expect when parsing an *anchor record*. Secondly the sorting of updates and the inclusion of a bloom filter allows indexing services to effectively determine if an update was made which is relevant, and find this update in the tree effectively.
+
+
+## Specification
+
+First lets define the basic IPLD data structures used.
+```ipldsch
+type MerkleNode [Link]
+
+type TreeMetaData struct {
+  numEntries Int
+  bloomType String
```

Comment:
I wonder if we want more than just a string for specifying the bloomType here.  In order to consume the bloom filter users will have to know the arguments used to construct it (either size of the bloom filter and number of hash functions, or target number of items and false positive rate).  To provide that to them we'll either need a public mapping of the `bloomType` string to the arguments used to construct it (so that we could change the `bloomType` string if we ever wanted to change the arguments we're using to construct the bloom filter, even if the underlying bloom filter library we're using doesn't change) or we need to encode those arguments here.  I think putting them here probably makes more sense

Reply:
Actually we have to put it here, because there's no way to know how many entries are in the bloom filter otherwise.  Different anchor merkle trees could wind up with a different number of filter entries, and if we're generating the bloom filter based on the exact known number of entries, then that's going to change the size of the filter.

Reply:
The library we intend on using will export to a format which includes this information: https://github.com/Callidon/bloom-filters#export-and-import
So the info will be in the `bloomFilter` property.